File: problems/famous_algorithms/kadanes_algorithm/js/kadanes_algorithm.js
export const kadanesAlgorithm = (array) => {
  let maxSoFar = Number.MIN_SAFE_INTEGER;
  let maxEndingHere = 0;

  for (let i = 0; i < array.length; i++) {
    maxEndingHere += array[i];

    maxSoFar = Math.max(maxSoFar, maxEndingHere);

    if (maxEndingHere < 0) maxEndingHere = 0;
  }
  return maxSoFar;
};

File: problems/famous_algorithms/kadanes_algorithm/cpp/kadanes_algorithm.cpp
#include "kadanes_algorithm.h"

int kadanesAlgorithm(std::vector<int> array) {
  int max_so_far = INT_MIN, max_ending_here = 0;

  for (auto it = array.begin(); it != array.end(); it++) {
    max_ending_here += *it;
    max_so_far = std::max(max_ending_here, max_so_far);

    if (max_ending_here < 0)
      max_ending_here = 0;
  }

  return max_so_far;
}
